### DISCUSSION TITLE
Customize the appearance of hover messages

### DISCUSSION THREAD

**Discussion by GSri30** [TypeFox/monaco-languageclient#509] [Q&A] [answered]
I was trying to customize the appearance of hover messages in monaco-editor. I am using `commonmark` package to customize the messages on the server side. Then I'm sending the rpc message via web-sockets to the monaco-editor. I have already installed the required language client services. 
Below are the examples of rpc-messages that are getting communicated:
```
Server receives from editor: {"jsonrpc":"2.0","id":4,"method":"textDocument/hover","params":{"textDocument":{"uri":"inmemory:/demo/240e9ff8-2b3c-8cd2-61b0-18fb9c619bfb.tsl"},"position":{"line":0,"character":35}}}

Server sends to editor: {"jsonrpc":"2.0","id":4,"result":{"contents":{"kind":"markdown","value":"\u003cp\u003e\u003cstrong\u003eGUID_ELEMENT\u003c/strong\u003e\u003c/p\u003e\n  \n*GUID* element is a unique identifier for the entity."},"range":{"start":{"line":0,"character":0},"end":{"line":0,"character":42}}}}
```
Here, server is sending the message in html format to the client. But I'm not able to get the customization to my hover message on the client side, i.e. language client isn't able to parse the html. Can anyone help me out.

**Comment by GSri30** [comment]
For your reference, the following is the code snippet on the server side

```
public class CustomHoverService extends HoverService {
    @Inject
    private CustomIEObjectDocumentationProvider eObjectDocumentationProvider;

    @Inject
    private INameLabelProvider nameLabelProvider;

    @Override
    public String getContents(EObject element) {
        String documentation = eObjectDocumentationProvider.getDocumentation(element);
        if (documentation == null) {
            return formatHoverContent(getFirstLine(element));
        } else {
            return formatHoverContent(getFirstLine(element) + "  \n" + documentation);
        }
    }

    public String getFirstLine(EObject o) {
        String label = nameLabelProvider.getNameLabel(o);
        return "**" + o.eClass().getName() + "**" + (label != null ? " *" + label + "*" : "");
    }

    private String formatHoverContent(String content) {
        Parser parser = Parser.builder().build();
        Node document = parser.parse(content);
        HtmlRenderer renderer = HtmlRenderer.builder().build();
        return renderer.render(document);
    }
}
```

**Comment by msujew** [comment]
Hey @GSri30,

> But I'm not able to get the customization to my hover message on the client side, i.e. language client isn't able to parse the html. Can anyone help me out.

The LSP [explicitly says that any response is supposed to be sanitized on client side](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover):

> Note that markdown strings will be sanitized - that means html will be escaped.

Monaco does exactly that when it receives a hover response. Otherwise, responses like `<script>alert('hello')</script>` could become quite problematic. I'm wondering though why you want to render your markdown to HTML yourself - if you send a [`MarkedString`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#markedString) back, it will be rendered as HTML on the frontend side as expected. In Xtext, it should be enough to return the markdown string from `getContents`.

**Reply by GSri30** [reply]
I just wanted to introduce colors to a few words to customize the hover message properly. From markdown, I can't achieve that.

**Reply by CGNonofr** [reply]
depending on what you need colors, you can still render a code block and do the coloring client side

**Reply by GSri30** [reply]
Can you provide examples, if there are any?

**Reply by CGNonofr** [reply, marked as answer]
Just a raw markdown code block =>
````
```myCustomLanguage
blablabla
```
````

and you register the myCustomLanguage language/grammar on monaco

**Reply by GSri30** [reply]
Oh yeah, got it. Thanks!
